### PR TITLE
Login authorisation and signup post.

### PR DESCRIPTION
**Pull request by JeldorPKU**
The interaction between the front-end and the back-end for login is now complete.

### CODE REVIEW

File: assets/js/race.js [hunk 1]
```diff
@@ -28,6 +28,7 @@ function removeIndividual(item) {
 }
 
 /*
+<<<<<<< HEAD
```
Comment (huxuan) [inline assets/js/race.js]: This is a conflict sign, please remove it (every time after you resolve the conflicts).

**Comment by JeldorPKU** [comment]
The email activation is removed due to the following errors.
```
<div style="border:1px solid #990000;padding-left:20px;margin:0 0 10px 0;">

<h4>A PHP Error was encountered</h4>

<p>Severity: Warning</p>
<p>Message:  fsockopen(): unable to connect to smtp.126.com:25 (Operation timed out)</p>
<p>Filename: libraries/Email.php</p>
<p>Line Number: 1689</p>

</div><div style="border:1px solid #990000;padding-left:20px;margin:0 0 10px 0;">

<h4>A PHP Error was encountered</h4>

<p>Severity: Warning</p>
<p>Message:  fwrite() expects parameter 1 to be resource, boolean given</p>
<p>Filename: libraries/Email.php</p>
<p>Line Number: 1846</p>
```
I will write an issue and fix it later. A more robust front-end validation will be followed up,

**Comment by JeldorPKU** [comment]
@huxuan The problem was that your VPN is not able to connect `smtp.126.com`. It is ok for a server with stable connection to the SMTP server.

File: application/config/constants.php [hunk 2]
```diff
@@ -82,5 +82,19 @@
...
 
+/*
+ * Error message settings;
+ * =======================
+ * These messages are error messages returned to user input.
+ *
+ */
+$ERR_MSG = array(
+    '200' => 'OK',
+    '202' => 'User not confirmed.',
```
Comment (huxuan) [inline application/config/constants.php]: We need to judge here. One option is to add period at the end of EVERY message, or remove them all.

File: application/controllers/user.php [hunk 3]
```diff
@@ -18,31 +18,45 @@ public function login() {
...
-            $this->session->set_userdata('logged_in', true);
-            $this->session->set_userdata('school_id', $user_info['school_id']);
-            redirect(site_url('registration/index'));
-        } else {
-            $this->load->view('login_form');
+            if ($user_info == NULL) {
+                $err_code = '204';
+            }
+
+            $data = array(
```
Comment (huxuan) [inline application/controllers/user.php]: I prefer we can have a global library here to send message.

Reference: http://stackoverflow.com/questions/9237112/code-igniter-global-function

File: application/controllers/user.php [hunk 4]
```diff
@@ -18,31 +18,45 @@ public function login() {
...
+        }
 
         if ($this->input->server('REQUEST_METHOD') == 'POST') {
             $login_info = $this->input->post();
+            header('Content-Type: application/json');
 
             if ($this->form_validation->run('login') == false) {
-                redirect(site_url('user/login'));
+                $err_code = '400';
+                $data = array(
```
Comment (huxuan) [inline application/controllers/user.php]: Dunplicate code snippet here. Please see the following comment.

File: application/controllers/user.php [hunk 5]
```diff
@@ -51,15 +65,32 @@ public function signup() {
...
         if ($this->input->server('REQUEST_METHOD') == 'POST') {
             $data = $this->input->post();
+            header('Content-Type: application/json');
+            $err_code = '200';
+
 
             if ($this->form_validation->run('signup') == false) {
-                redirect(site_url('user/signup'));
+                $err_code = '400';
+                $response = array(
```
Comment (huxuan) [inline application/controllers/user.php]: Dunplicate code snippet.

File: application/controllers/user.php [hunk 6]
```diff
@@ -51,15 +65,32 @@ public function signup() {
...
+                    'code' => $err_code,
+                    'msg' => $GLOBALS['ERR_MSG'][$err_code]
+                );
+                echo json_encode($response);
+                exit;
             }
+            $response = array(
```
Comment (huxuan) [inline application/controllers/user.php]: Dunplicate code snippet.

File: application/views/login_form.php [hunk 7]
```diff
@@ -1,15 +1,20 @@
...
+    var registration = "<?=site_url('registration')?>"
+</script>
+
 <div class="signcontainer">
     <h3>高校用户登录</h3>
     <hr/>
     <div class="col-sm-4">用户名</div>
     <div class="col-sm-8">
-        <input type="text" class="form-control" name="mail">
+        <input type="text" class="form-control" name="mail" id="mail">
```
Comment (huxuan) [inline application/views/login_form.php]: I think the only thing matters is `name` not `id`. Can you explain why you add this?

File: assets/js/race.js [hunk 8]
```diff
@@ -46,5 +46,116 @@ function removeTeam(item) {
     }
     refreshOrder();
 }
+/*
```
Comment (huxuan) [inline assets/js/race.js]: Add an empty line before the section.

File: assets/js/race.js [hunk 9]
```diff
@@ -46,5 +46,116 @@ function removeTeam(item) {
...
+        province: province,
+        leader: leader,
+        tel: tel,
+        mail: mail,
+        password: password,
+        passconf: passconf
+    };
+    $.post(controller, data, function(data) {
+        if (data.code == "200") {
+            alert("注册成功！请登录你的邮箱查看激活邮件！")
```
Comment (huxuan) [inline assets/js/race.js]: Prefer to use `您` instead of `你`?

File: application/views/login_form.php (hunk 7 above)
Comment (JeldorPKU) [inline application/views/login_form.php]: I just wanted to use the jQuery selector more conveniently. I didn't know how to select `name` then so I wrote the code like this. I can refer to the docs and improve it if it really matters.

File: assets/js/race.js (hunk 9 above)
Comment (JeldorPKU) [inline assets/js/race.js]: Get it.

File: application/config/constants.php (hunk 2 above)
Comment (JeldorPKU) [inline application/config/constants.php]: Sorry it was my mistake.

File: application/controllers/user.php (hunk 3 above)
Comment (JeldorPKU) [inline application/controllers/user.php]: I am not sure I correctly understand you. I think the problem here is to make `registration/***` a private page which only logged in users can see. And this issue may be #14 . This is the only controller which can be seen only when the user is logged in.

File: application/views/login_form.php (hunk 7 above)
Comment (huxuan) [inline application/views/login_form.php]: OK, I got what you mean.

**Pull request merged**